hazelcast/src/main/java/com/hazelcast/internal/networking/nio/NioOutboundPipeline.java

```diff
@@ -206,73 +286,116 @@ public void process() throws Exception {
 
         flushToSocket();
 
+        if (migrationRequested()) {
+            startMigration();
+            // we leave this method and the NioOutboundPipeline remains scheduled.
+            // So we don't need to worry about write-through
+            return;
+        }
+
         if (sendBuffer.remaining() > 0) {
             pipelineStatus = DIRTY;
         }
 
         switch (pipelineStatus) {
             case CLEAN:
-                // There is nothing left to be done; so lets unschedule this pipeline
-                unschedule();
+                postProcessClean();
                 break;
             case DIRTY:
-                // pipeline is dirty, so lets register for an OP_WRITE to write
-                // more data.
-                registerOp(OP_WRITE);
+                postProcessDirty();
                 break;
             case BLOCKED:
-                // pipeline is blocked; no point in receiving OP_WRITE events.
-                unregisterOp(OP_WRITE);
+                postProcessBlocked();
                 break;
             default:
                 throw new IllegalStateException();
         }
     }
 
-    /**
-     * Tries to unschedule this pipeline.
-     * <p>
-     * It will only be unscheduled if:
-     * - there are no pending frames.
-     * <p>
-     * If the outputBuffer is dirty then it will register itself for an OP_WRITE since we are interested in knowing
-     * if there is more space in the socket output buffer.
-     * If the outputBuffer is not dirty, then it will unregister itself from an OP_WRITE since it isn't interested
-     * in space in the socket outputBuffer.
-     * <p>
-     * This call is only made by the owning IO thread.
-     */
-    private void unschedule() throws IOException {
-        // since everything is written, we are not interested anymore in write-events, so lets unsubscribe
+    private void postProcessBlocked() throws IOException {
+        // pipeline is blocked; no point in receiving OP_WRITE events.
         unregisterOp(OP_WRITE);
-        // So the outputBuffer is empty, so we are going to unschedule the pipeline.
-        scheduled.set(false);
 
-        if (writeQueue.isEmpty() && priorityWriteQueue.isEmpty()) {
-            // there are no remaining frames, so we are done.
-            return;
+        // we try to set the state to blocked.
+        for (; ; ) {
+            State state = scheduled.get();
+            if (state == State.SCHEDULED) {
+                // if it is still scheduled, we'll just try to put it to blocked.
+                if (scheduled.compareAndSet(State.SCHEDULED, State.BLOCKED)) {
+                    break;
+                }
+            } else if (state == State.BLOCKED) {
+                // it is already blocked, so we are done.
+                break;
+            } else if (state == State.RESCHEDULE) {
+                // rescheduling is requested, so lets do that. Once put to RESCHEDULE,
```

Comment:
`only the IO thread will change the state` -> is this still true in the case of write-through?
Can we have two threads running this method where one thread sets the state to `SCHEDULED` after which the other thread sets the state to `BLOCKED`?

Reply:
The documentation is not valid. The only thread running this method is the one that 'owns' the pipeline; and that is at most one. It could be the IO thread, could be another thread in case of write through.

So you can't have 2 threads running this method.